app/addons/documents/changes/stores.js

```diff
@@ -15,51 +15,58 @@ define([
   'addons/documents/changes/actiontypes'
 ], function (FauxtonAPI, ActionTypes) {
 
-  var Stores = {};
-
 
-  // tracks the state of the header (open/closed)
-  var ChangesHeaderStore = FauxtonAPI.Store.extend({
+  var ChangesStore = FauxtonAPI.Store.extend({
     initialize: function () {
       this.reset();
     },
 
     reset: function () {
       this._tabVisible = false;
+      this._filters = [];
+      this._changes = [];
+      this._databaseName = '';
+      this._maxChangesListed = 1000;
+      this._showingSubset = false;
     },
 
-    toggleTabVisibility: function () {
-      this._tabVisible = !this._tabVisible;
+    setChanges: function (options) {
+      this._filters = options.filters;
+      this._databaseName = options.databaseName;
+      this._changes = _.map(options.changes, function (change) {
+        return {
+          id: change.get('id'),
+          seq: change.get('seq'),
+          deleted: change.get('deleted') ? change.get('deleted') : false,
+          changes: change.get('changes'),
+          doc: change.get('doc') // only populated with ?include_docs=true
+        };
+      });
     },
 
-    isTabVisible: function () {
-      return this._tabVisible;
+    getChanges: function () {
+      var matches = _.reduce(this._filters, function (elements, filter) {
+        return _.filter(elements, function (element) {
+          var match = false;
+          _.each(element, function (value) {
+            if (!_.isUndefined(value) && new RegExp(filter, 'i').test(value.toString())) {
+              match = true;
+            }
+          });
+          return match;
+        });
+      }, this._changes, this);
+
+      this._showingSubset = matches.length > this._maxChangesListed;
+      return matches.slice(0, this._maxChangesListed);
     },
```

Comment:
you can already limit the max-changes before the reduceby slicing them - that should increase performance a lot